Example for using with Mapbox GL JS

xiaxiangfeng opened the pull request:
add a Example for using with Mapbox GL JS #426

gkjohnson reviewed (commented):
Thank you! I've added a few notes - and do you mind adding comments to the code? Especially the "mapboxExampleCamera". It's not easily clear what it's for or what it does from looking at the code. Is it the case that it needs to be a class like this?

xiaxiangfeng commented on example/mapboxExampleCamera.js:
```diff
@@ -0,0 +1,287 @@
+			: 0;
+
+		const cameraToSeaLevelDistance =
+      ( t._camera.position[ 2 ] * worldSize - minElevationInPixels ) /
+      Math.cos( t._pitch );
+
+		const topHalfSurfaceDistance =
+      ( Math.sin( fovAboveCenter ) * cameraToSeaLevelDistance ) /
+      Math.sin(
+      	utils.clamp(
+      		Math.PI - groundAngle - fovAboveCenter,
+      		0.01,
+      		Math.PI - 0.01
+      	)
+      );
```
ok

xiaxiangfeng commented on example/mapboxExample.js:
```diff
@@ -0,0 +1,145 @@
+
+const origin = [ 116, 35 ];
+
+// TO MAKE THE MAP APPEAR YOU MUST
+// ADD YOUR ACCESS TOKEN FROM
+// https://account.mapbox.com
+// This is my private key, it is prohibited to use in the project
+window.mapboxgl.accessToken =
+        'pk.eyJ1IjoieGlheGlhbmdmbmVnIiwiYSI6ImNscTRzc245NTA5c3cya3BhdXA4MDY0NWQifQ._9T_J9z7Hvcd00nQL3UuSw';
+const map = new window.mapboxgl.Map( {
```
This is a key created by my own account, which can be used normally. I just don't want anyone to use my key in their project. If someone uses it, they will have to bear the risk of failure themselves

xiaxiangfeng replied:
This comment can be deleted

// This is my private key, it is promoted to use in the project

gkjohnson replied:
I think it's best to use the same model as the other examples for the moment and require that users provide their own key in the GUI to use the demo. Otherwise it's difficult to track who is responsible for which key and where they've come from.

xiaxiangfeng replied:
ok

xiaxiangfeng commented on example/mapboxExample.js:
```diff
@@ -0,0 +1,145 @@
+
+			new CameraSync( map, camera, world );
+
+			const url = 'https://xiaxiangfeng.github.io/3DTilesRendererJS.Test/baowei21/tileset.json';
+
+			tiles = new TilesRenderer( url );
```
This is the tileset I generated using the tool, and all the b3DM files are placed in it https://xiaxiangfeng.github.io/3DTilesRendererJS.Test/ Under the directory, it can be downloaded

xiaxiangfeng replied:
Alternatively, you can modify it to another tileset

gkjohnson replied:
What is the original source of the model, though? I would like to confirm the license and data used for the model.

xiaxiangfeng replied:
https://raw.githubusercontent.com/NASA-AMMOS/3DTilesSampleData/master/msl-dingo-gap/0528_0260184_to_s64o256_colorize/0528_0260184_to_s64o256_colorize/0528_0260184_to_s64o256_colorize_tileset.json

xiaxiangfeng commented:
> Thank you! I've added a few notes - and do you mind adding comments to the code? Especially the "mapboxExampleCamera". It's not easily clear what it's for or what it does from looking at the code. Is it the case that it needs to be a class like this?

You can add comments or modify some code

xiaxiangfeng commented:
> > Thank you! I've added a few notes - and do you mind adding comments to the code? Especially the "mapboxExampleCamera". It's not easily clear what it's for or what it does from looking at the code. Is it the case that it needs to be a class like this?
> 
> You can add comments or modify some code

The absence of comments in mapboxExampleCamera does make it difficult for people to understand. I can add some comments to make it easier to understand

xiaxiangfeng commented:
modify：1.users provide their own key in the GUI to use the demo

xiaxiangfeng commented:
modify：
1.users provide their own key in the GUI to use the demo
2.changed tileset url
3.Add comments to the core of the code in the mapboxExampleCamera file

gkjohnson commented on example/mapboxExample.js:
```diff
@@ -0,0 +1,176 @@
+
+const params = {
+
+	'accessToken': 'pk.eyJ1IjoieGlheGlhbmdmbmVnIiwiYSI6ImNscTRzc245NTA5c3cya3BhdXA4MDY0NWQifQ._9T_J9z7Hvcd00nQL3UuSw',
```
It looks like we're still providing a token? I think we should default to a string called "put-your-api-key-here" as the Google Maps example does.

xiaxiangfeng replied:
ok

gkjohnson commented on example/mapboxExample.js:
```diff
@@ -0,0 +1,176 @@
+				scene.add( world );
+
+				// lights
+				const dirLight = new THREE.DirectionalLight( 0xffffff, 0.8 );
+				dirLight.position.set( 1, 2, 3 );
+				world.add( dirLight );
+
+				const ambLight = new THREE.AmbientLight( 0xffffff, 1 );
+				world.add( ambLight );
```
The tileset used is using basic materials which don't use lighting - I think we can remove these for now.

xiaxiangfeng replied:
ok

gkjohnson commented on example/mapboxExampleCamera.js:
```diff
@@ -0,0 +1,284 @@
+
+// 512 * 2000
+// 512: TILE_SIZE
+const WORLD_SIZE = 10240000;
+const FOV_ORTHO = ( 0.1 / 180 ) * Math.PI; // Mapbox doesn't accept 0 as FOV
+const FOV = Math.atan( 3 / 4 ); // from Mapbox https://github.com/mapbox/mapbox-gl-js/blob/main/src/geo/transform.js#L93
+const EARTH_RADIUS = 6371008.8; // from Mapbox  https://github.com/mapbox/mapbox-gl-js/blob/0063cbd10a97218fb6a0f64c99bf18609b918f4c/src/geo/lng_lat.js#L11
+const EARTH_CIRCUMFERENCE_EQUATOR = 40075017; // from Mapbox https://github.com/mapbox/mapbox-gl-js/blob/0063cbd10a97218fb6a0f64c99bf18609b918f4c/src/geo/lng_lat.js#L117
+const MapConstants = {
+	WORLD_SIZE: WORLD_SIZE,
+	PROJECTION_WORLD_SIZE: WORLD_SIZE / ( EARTH_RADIUS * Math.PI * 2 ),
+	MERCATOR_A: EARTH_RADIUS,
+	DEG2RAD: Math.PI / 180,
+	RAD2DEG: 180 / Math.PI,
+	EARTH_RADIUS: EARTH_RADIUS,
+	EARTH_CIRCUMFERENCE: 2 * Math.PI * EARTH_RADIUS,
+	EARTH_CIRCUMFERENCE_EQUATOR: EARTH_CIRCUMFERENCE_EQUATOR,
+	FOV_ORTHO: FOV_ORTHO,
+	FOV: FOV,
+	FOV_DEGREES: ( FOV * 180 ) / Math.PI,
+	TILE_SIZE: 512,
+};
```
It looks like a number of these constants aren't used in the demo - can we remove the unnecessary ones?

xiaxiangfeng replied:
ok

gkjohnson commented on example/mapboxExample.js:
```diff
@@ -0,0 +1,176 @@
+}
+
+let map;
+const origin = [ 116, 35 ];
```
What are the units fore these values? Are these lat / lon values?

xiaxiangfeng replied:
Yes, it's longitude and latitude

gkjohnson commented on example/mapboxExampleCamera.js:
```diff
@@ -0,0 +1,284 @@
+import * as THREE from 'three';
+
+const utils = {
+	clamp: function ( n, min, max ) {
+
+		return Math.min( max, Math.max( min, n ) );
+
+	},
```
This function is redundant to three.js' `MathUtils.clamp` function

xiaxiangfeng replied:
ok

gkjohnson commented on example/mapboxExampleCamera.js:
```diff
@@ -0,0 +1,284 @@
+	TILE_SIZE: 512,
+};
+
+const Constants = MapConstants;
```
This redefinition of the constants object seems unnecessary?

gkjohnson commented on example/mapboxExampleCamera.js:
```diff
@@ -0,0 +1,284 @@
+		// Postion and configure the world group so we can scale it appropriately when the camera zooms
+		this.world = world || new THREE.Group();
+		this.world.position.x = this.world.position.y =
+    MapConstants.WORLD_SIZE / 2;
```
The rationale for some of these changes should be documented, I think. What's the reasoning for setting the position here? We disable matrix auto update and set the matrix explicitly later so this will never really be used, right?

gkjohnson commented on example/mapboxExampleCamera.js:
```diff
@@ -0,0 +1,284 @@
+			worldSize: MapConstants.TILE_SIZE * this.map.transform.scale,
+		};
+		// Listen for move events from the map and update the Three.js camera
+		const _this = this;
+		this.map
+			.on( 'move', function () {
+
+				_this.updateCamera();
+
+			} )
+			.on( 'resize', function () {
+
+				_this.setupCamera();
+
+			} );
```
If we use arrow functions there's no need to declare a local scope "_this". ie

```js
this.map
	.on( 'move', () => {

		this.updateCamera();

	} )
```

xiaxiangfeng replied:
ok

gkjohnson commented on example/mapboxExampleCamera.js:
```diff
@@ -0,0 +1,284 @@
+
+	}
+
+	circumferenceAtLatitude( latitude ) {
+
+		return (
+			MapConstants.EARTH_CIRCUMFERENCE *
+      Math.cos( ( latitude * Math.PI ) / 180 )
+		);
+
+	}
```
This function looks like it's not used?

xiaxiangfeng replied:
Used in the mergetorZfromAltitude function

gkjohnson commented on example/mapboxExampleCamera.js:
```diff
@@ -0,0 +1,284 @@
+
+		const h = t.height;
+		const w = t.width;
+		this.camera.projectionMatrix = utils.makePerspectiveMatrix(
+			t._fov,
+			w / h,
+			nearZ,
+			farZ
+		);
```
There shouldn't be a need to manually construct a perspective matrix. These are all values that can be set on the camera itself and "updateProjectionMatrix" can be called to correctly update the projection matrix and ensure the inverse projection matrix are set

xiaxiangfeng replied:
Unlike the Mapbox GL JS camera, separate camera translation and rotation out into its world matrix If this is applied directly to the projection matrix, it will work OK but break raycasting

gkjohnson replied:
I'm not following. I don't think we should include position or rotation values in the projection matrix. With the values you have here you can set the three.js camera parameters like so:

```js
camera.fov = t._fov;
camera.aspect = w / h;
camera.near = nearZ;
camera.far = farZ;
camera.updateProjectionMatrix();
```

xiaxiangfeng replied:
We need to use the same projectionMatrix as mapboxgl. I modified this place and used the method this.camera.projectionMatrix.elements = t._camera.getCameraToClipPerspective provided by mapboxgl. I found that the calculation methods of projectionMatrix of mapboxgl and threejs are somewhat different. Mapboxgl uses the gl-matrix library.

gkjohnson replied:
I've taken a look at the gl-matrix code [here](https://glmatrix.net/docs/mat4.js.html#line1474) and the three.js projection matrix code [here]() and I don't see any difference in calculations beyond superficial organization. Can you explain what the differences you saw are?

gkjohnson commented on example/mapboxExampleCamera.js:
```diff
@@ -0,0 +1,284 @@
+		pixelsPerMeter = this.mercatorZfromAltitude( 1, t.center.lat ) * worldSize;
+		const fovAboveCenter = t._fov * ( 0.5 + t.centerOffset.y / t.height );
+
+		// Adjust distance to MSL by the minimum possible elevation visible on screen,
+		// this way the far plane is pushed further in the case of negative elevation.
```
I assume MSL means "mean sea level" here?

xiaxiangfeng replied:
yes

gkjohnson commented on example/mapboxExampleCamera.js:
```diff
@@ -0,0 +1,284 @@
+
+		// Adjust distance to MSL by the minimum possible elevation visible on screen,
+		// this way the far plane is pushed further in the case of negative elevation.
+		const minElevationInPixels = t.elevation
+			? t.elevation.getMinElevationBelowMSL() * pixelsPerMeter
+			: 0;
```
How can we compute the elevation in pixels? This is a non constant value since we're using a perspective camera. 1 meter far from the camera will be represented by fewer pixels than a meter closer to the camera.

xiaxiangfeng replied:
Our purpose is to calculate farZ, which all happens before projection

gkjohnson replied:
I don't understand. How can you calculate a distance in pixels if it's not relative to the screen?

xiaxiangfeng replied:
I can't seem to answer your question. These codes refer to the source code of mapboxgl. I have modified this place now. farZ directly uses the one provided by mapboxgl.

xiaxiangfeng replied:
> pixels

The pixels here are the sizes under the Mercator projection, which is just a possible value calculated.

xiaxiangfeng commented:
> Thanks for adding some comments! I've written a few more notes. I think the "updateCamera" portion of the code should be able to be simplified quite a bit. It would be best to use as many built in three.js functions to compute the position and camera matrices as possible so it's easier to follow. Accessing direct matrix elements and keeping copies of different matrices around makes the code more difficult to follow. As does keeping some functions around that are unused.
> 
> I'm happy to clean things up a bit afterward but I just want to make sure I understand what's happening since I'm new to MapBox and it's clear to others, as well.
> 
> Also a couple comments about the demo:
> 
> * It looks like the camera world matrix is being updated a frame late which is causing pieces of the tileset to load in delayed which causes a flash. We may need to call "camera.updateMatrixWorld" before triggering "tiles.update()" in the render function.
> * When zooming into the 3d tiles model from above parts of the model disappears - it seems like something isn't being updated correctly:
> 
> <img alt="image" width="1361" src="https://private-user-images.githubusercontent.com/734200/294158962-373d8f5b-7bd5-4d75-b435-fe19c8b57245.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ0MjA1NjksIm5iZiI6MTcwNDQyMDI2OSwicGF0aCI6Ii83MzQyMDAvMjk0MTU4OTYyLTM3M2Q4ZjViLTdiZDUtNGQ3NS1iNDM1LWZlMTljOGI1NzI0NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwNVQwMjA0MjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hYjY1ZGJhODYzNDJmMThlMmIyNzc4MmVhNzU0ZjM0YzMxZjE5OTc3MDE0YWMzNDg1N2VjYTJlZTE5ZGJiYTFiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.PSvCbuTvOD7-vDJrRA5vKcbdVNn-igMx8zCTmoLqegM">

1. I added camera.updateMatrixWorld but nothing seems to change
2. The tileset used now has terrain, it needs to be placed in its correct position, and mapboxgl needs to be turned on for terrain. We can also set tilesGroup.position.z = 1; make the tileset higher than the mapboxgl ground

gkjohnson commented:
> I added camera.updateMatrixWorld but nothing seems to change

Interesting - once the code is simplified more I can take a look.

> The tileset used now has terrain, it needs to be placed in its correct position, and mapboxgl needs to be turned on for terrain. We can also set tilesGroup.position.z = 1; make the tileset higher than the mapboxgl ground

Does this explain why the map plane seems to be moving, though? As you zoom in and out and rotate the camera the portion of the terrain being hidden seems to change.

xiaxiangfeng commented:
> > I added camera.updateMatrixWorld but nothing seems to change
> 
> Interesting - once the code is simplified more I can take a look.
> 
> > The tileset used now has terrain, it needs to be placed in its correct position, and mapboxgl needs to be turned on for terrain. We can also set tilesGroup.position.z = 1; make the tileset higher than the mapboxgl ground
> 
> Does this explain why the map plane seems to be moving, though? As you zoom in and out and rotate the camera the portion of the terrain being hidden seems to change.

When z is increased, there is still a problem, it seems to be related to depth writing, I am not sure

xiaxiangfeng commented:
modify：
1. Delete the default accessToken
2. Add camera.updateMatrixWorld
3. Simplify the code of mapboxExampleCamera
4.Delete lights
5.Set tilesGroup.position.z = 1

gkjohnson commented:
> When z is increased, there is still a problem, it seems to be related to depth writing, I am not sure

I don't this can be merged until it works correctly and I don't have the bandwidth to debug these issues. I recommend asking at the three.js forum to find help with how to properly integrate three.js with mapbox and extract the camera transform and camera frustum separately. Alternatively we could update the tiles renderer project to be resilient to this kind of matrix but I'd prefer to avoid that and this doesn't fix the raycasting issue. Ideally mapbox would provide a separate view and projection matrix (there was a discussion [here](https://github.com/mapbox/mapbox-gl-js/issues/7395) but it didn't go anywhere).

For reference it looks like there's MapBoxGL documentation for three.js integration [here](https://docs.mapbox.com/mapbox-gl-js/example/add-3d-model/) and [threebox](https://github.com/jscastro76/threebox/tree/master) provides some integration utilities.

---

Relating to threebox - it looks like you have have copy and pasted quite a bit of code from the threebox project without attribution. At least from the CameraSync file and other constants. Is this the case? Its your responsibility to properly attribute code and notify projects of where the code has come from if it is not yours.

xiaxiangfeng commented:
> > When z is increased, there is still a problem, it seems to be related to depth writing, I am not sure
> 
> I don't this can be merged until it works correctly and I don't have the bandwidth to debug these issues. I recommend asking at the three.js forum to find help with how to properly integrate three.js with mapbox and extract the camera transform and camera frustum separately. Alternatively we could update the tiles renderer project to be resilient to this kind of matrix but I'd prefer to avoid that and this doesn't fix the raycasting issue. Ideally mapbox would provide a separate view and projection matrix (there was a discussion [here](https://github.com/mapbox/mapbox-gl-js/issues/7395) but it didn't go anywhere).
> 
> For reference it looks like there's MapBoxGL documentation for three.js integration [here](https://docs.mapbox.com/mapbox-gl-js/example/add-3d-model/) and [threebox](https://github.com/jscastro76/threebox/tree/master) provides some integration utilities.
> 
> Relating to threebox - it looks like you have have copy and pasted quite a bit of code from the threebox project without attribution. At least from the CameraSync file and other constants. Is this the case? Its your responsibility to properly attribute code and notify projects of where the code has come from if it is not yours.

The code in this file comes from https://github.com/jscastro76/threebox/tree/master I have simplified and modified some content

xiaxiangfeng commented:
First of all, I would like to thank gkjohnson for his support these days. I will close this PR for now. I feel that it cannot solve the existing problems at the moment. I also hope that someone who encounters a similar situation in the future can provide help or complete this PR.
